src/femtomes.cpp

```diff
@@ -189,116 +286,213 @@ int GPSDriverFemto::receive(unsigned timeout)
-						   (uint8_t *)&_femto_msg.header.data, (uint32_t)0);
-		crc = calculateBlockCRC32((uint32_t)_femto_msg.header.femto_header.messagelength, (uint8_t *)&_femto_msg.data[0], crc);
+	}else{	/**< RTCM mode */
 
-		if (_femto_msg.crc == crc) {
-			iRet = _femto_msg.read;
+        switch (_decode_state)
+        {
+		case FemtoDecodeState::pream_ble1:
+			if(temp == '$')
+			{
+				_decode_state = FemtoDecodeState::pream_nmea_got_sync1;
+				_femto_msg.read = 0;
+				_femto_msg.data[_femto_msg.read++] = temp;
+			}
+			else if(temp == RTCM3_PREAMBLE && _rtcm_parsing)
+			{
+				_decode_state = FemtoDecodeState::decode_rtcm3;
+				_rtcm_parsing->addByte(temp);
+			}
+			break;
+		case FemtoDecodeState::pream_nmea_got_sync1:
+			if(temp == '$')
+			{
+				_decode_state = FemtoDecodeState::pream_nmea_got_sync1;
+				_femto_msg.read = 0;
+			}
+			else if(temp == '*')
+			{
+				_decode_state = FemtoDecodeState::pream_nmea_got_asteriks;
+			}
+			else if(temp == RTCM3_PREAMBLE && _rtcm_parsing)
+			{
+				_decode_state = FemtoDecodeState::decode_rtcm3;
+				_rtcm_parsing->addByte(temp);
+			}
 
-		} else {
-			FEMTO_DEBUG("Femto: data packet is bad");
+			if (_femto_msg.read >= (sizeof(_femto_msg.data) - 5))
+			{
+				FEMTO_DEBUG("buffer overflow")
+				_decode_state = FemtoDecodeState::pream_ble1;
+				_femto_msg.read = 0;
+			}
+			else
+			{
+				_femto_msg.data[_femto_msg.read++] = temp;
+			}
+			break;
+		case FemtoDecodeState::pream_nmea_got_asteriks:
+			_femto_msg.data[_femto_msg.read++] = temp;
+			_decode_state = FemtoDecodeState::pream_nmea_got_first_cs_byte;
+			break;
+		case FemtoDecodeState::pream_nmea_got_first_cs_byte:
+		{
+			_femto_msg.data[_femto_msg.read++] = temp;
+			uint8_t checksum = 0;
+			uint8_t *buffer = _femto_msg.data + 1;
+			uint8_t *bufend = _femto_msg.data + _femto_msg.read - 3;
+
+			for (; buffer < bufend; buffer++)
+			{
+				checksum ^= *buffer;
+			}
+
+			if ((HEXDIGIT_CHAR(checksum >> 4) == *(_femto_msg.data + _femto_msg.read - 2)) &&
+				(HEXDIGIT_CHAR(checksum & 0x0F) == *(_femto_msg.data + _femto_msg.read - 1)))
+			{
+				iRet = _femto_msg.read;
+				_femto_msg.header.femto_header.messageid = FEMTO_MSG_ID_GPGGA;
+				FEMTO_DEBUG("Femto: got NMEA message with length %i",_femto_msg.read)
+			}
+			decodeInit();
+			break;
 		}
+		case FemtoDecodeState::decode_rtcm3:
+			if (_rtcm_parsing->addByte(temp))
+			{
+				FEMTO_DEBUG("Femto: got RTCM message with length %i", (int)_rtcm_parsing->messageLength())
+				_femto_msg.header.femto_header.messageid  = FEMTO_MSG_ID_RTCM3;
+				iRet = _rtcm_parsing->messageLength();
+			}
+			break;
+		default:
+			break;
+        }
 
-		break;
 	}
 
+
 	return iRet;
 }
 
 void GPSDriverFemto::decodeInit()
 {
 	_decode_state = FemtoDecodeState::pream_ble1;
-
+    /** init or reset rtcm parsing */
+    if(_output_mode == OutputMode::RTCM)
+    {
+        if(!_rtcm_parsing)
+        {
+            _rtcm_parsing = new RTCMParsing();
+        }
+        _rtcm_parsing->reset();
```

Comment:
```suggestion
        if(_rtcm_parsing) {
        _rtcm_parsing->reset();
        }
```

Reply:
Hi bkueng ! I don't understand this suggestion.

Reply:
The allocation might fail if out of memory (not a problem on the GCS, but on the vehicle), so we need to check again if not null.